File: TAS/Modules/Andreas/Python/InputReader.py
# InputReader Module
# @author: Andreas Theys
# @ version 6.0
# This module is the main processing unit for the data input stream.


# Imported modules 
# Basic math functionality
from math import *
# Sorting procedures of the datapoints
from Sort import *
# Processing of geographical data
from Coordinate import *
from FlightPath import *


# Filters dataset from 49 to 8 entities
# Provides correct data value conversions
# NOTE[1]: all type casting happens in this method
# NOTE[2]: massively important for further data handling
def filterData(datapoint):
    datapoint = list(datapoint)
    if len(datapoint) == 49:
        datapoint = datapoint[0:13]
        del datapoint[10]
        del datapoint[8]
        del datapoint[7]
        del datapoint[3]
        del datapoint[2]
        datapoint[0] = float(datapoint[0])                      # Time [s]
        datapoint[1] = str(datapoint[1]).replace(" ","")        # ID [] 
        datapoint[2] = (pi/180.)*float(datapoint[2])            # Latitude [rad]
        datapoint[3] = (pi/180.)*float(datapoint[3])            # Longitude [rad]
        datapoint[4] = 0.3048*float(datapoint[4])               # Altitude [m]
        temp1 = 0.51444444*float(datapoint[5])
        temp2 = (pi/180.)*float(datapoint[7])
        datapoint[5] = (0.3048/60.)*float(datapoint[6])         # Vx 
        datapoint[6] = temp1*sin(temp2)                         # Vy
        datapoint[7] = temp1*cos(temp2)                         # Vz
        # WSG84 coordinate transformation
        datapoint[2],datapoint[3],datapoint[4] = convertToXY2(datapoint[2],datapoint[3],datapoint[4])
        return datapoint


# Filters dataset from 49 to 8 entities
# Provides correct data value conversions
# NOTE[1]: all type casting happens in this method
# NOTE[2]: massively important for further data handling
def filterData2(datapoint):
    datapoint = list(datapoint)
    datapoint[0] = float(datapoint[0])      # Time [s]
    datapoint[1] = str(datapoint[1])        # ID [] 
    datapoint[2] = float(datapoint[2])      # Latitude [rad]
    datapoint[3] = float(datapoint[3])      # Longitude [rad]
    datapoint[4] = float(datapoint[4])      # Altitude [m]
    datapoint[5] = float(datapoint[5])      # Vx 
    datapoint[6] = float(datapoint[6])      # Vy
    datapoint[7] = float(datapoint[7])      # Vz
    return datapoint


# Processes line directly read in from datafile 
# Helper function
# NOTE[1]: function established for optimization purposes
def process(point,datapointsFixed):
        point = str(point)
        point = point.split(',')
        for value in point:
            value = str(value).replace(" ","")
        point = filterData(point) 
        datapointsFixed.append(point)
        return
        

# Processes line directly read in from datafile 
# Helper function
# NOTE[1]: function established for optimization purposes
def process2(point,datapointsFixed):
        point = str(point)
        point = point.split(',')
        for value in point:
            value = str(value).replace(" ","")
        point = filterData2(point) 
        datapointsFixed.append(point)
        return


# Main read-in method for datafiles
# Reads in data file and converts lines into filtered datapoints
def read(name):
    name = str(name)
    try:
        File = open(name,'r') 
    except ValueError:
        print "Wrong file!"
    datapoints = File.readlines()
    File.close()
    datapointsFixed = []
    map(lambda x: process(x,datapointsFixed),datapoints)
    IDs, times, sortedDPoints = sortByID(datapointsFixed)
    speeds(sortedDPoints)
    return IDs, times, sortedDPoints
    
    
# Main read-in method for datafiles
# Reads in data file and converts lines into filtered datapoints
def read2(name):
    name = str(name)
    try:
        File = open(name,'r') 
    except ValueError:
        print "Wrong file!"
    datapoints = File.readlines()
    File.close()
    datapointsFixed = []
    map(lambda x: process2(x,datapointsFixed),datapoints)
    IDs, times, sortedDPoints = sortByID2(datapointsFixed)
    return IDs, times, sortedDPoints

File: TAS/Modules/Andreas/Java/InputReader.java

import java.io.*;
import java.util.ArrayList;
import java.util.Set;
import java.util.HashSet;
import java.util.Collections;

/*******************************************
 * Class InputReader (static methods only)
 * @author Andreas Theys
 * @version 2.0
 *******************************************/
public abstract class InputReader{

	/*****************************************************
	 * Read-in method for datafile.
	 * @param name			name of file to read in.
	 * @return result		list of filtered datapoints.
	 * @throws IOException
	 *****************************************************/
	public static ArrayList<Datapoint> read(String name) throws IOException{
		ArrayList<Datapoint> result = new ArrayList<Datapoint>();
		BufferedReader br = new BufferedReader(new FileReader(name));
		br.readLine();
		String tempLine;
		while((tempLine = br.readLine()) != null){
			String[] tempSet = tempLine.split(",");
			for(int i=0; i<tempSet.length; i++){
				tempSet[i] = tempSet[i].replace(" ", "");
			}
			result.add(filter(tempSet));
		}
		br.close();
		return result;
	}
		
	/*****************************************************************
	 * filter-method.
	 * @param datapoint 	list of read in& separated Strign values
	 * @return 				filtered Datapoint-object.
	 *****************************************************************/
	public static Datapoint filter(String[] datapoint){
		double t = Double.parseDouble(datapoint[0]);
		String ID = datapoint[1];
		double x = Double.parseDouble(datapoint[2]);
		double y = Double.parseDouble(datapoint[3]);
		double z = Double.parseDouble(datapoint[4]);
		double vx = Double.parseDouble(datapoint[5]);
		double vy = Double.parseDouble(datapoint[6]);
		double vz = Double.parseDouble(datapoint[7]);
		return new Datapoint(t,ID,x,y,z,vx,vy,vz);
	}
	
	/***********************************************
	 * ID collection method.
	 * Collects all existing IDs in datafile. 
	 * @param points	list of Datapoint objects.
	 * @return result	list of Datapoint IDs.
	 ***********************************************/
	public static ArrayList<String> ids(ArrayList<Datapoint> points){
		ArrayList<String> result = new ArrayList<String>();
		for(Datapoint point: points){
			result.add(point.getID());
		}
		Set<String> hs = new HashSet<>();
		hs.addAll(result);
		result.clear();
		result.addAll(hs);
		return result;
	}
	
	
	/*************************************************
	 * Time collection method.
	 * Collects all existing time specs in datafile.
	 * @param points	list of Datapoint objects.
	 * @return result	list of times.
	 *************************************************/
	public static ArrayList<Double> times(ArrayList<Datapoint> points){
		ArrayList<Double> result = new ArrayList<Double>();
		for(Datapoint point: points){
			result.add(point.getT());
		}
		Set<Double> hs = new HashSet<>();
		hs.addAll(result);
		result.clear();
		result.addAll(hs);
		Collections.sort(result);
		return result;
	}
	
	
	/*************************************************
	 * Trajectory collection method.
	 * Collects all existing trajectories in datafile.
	 * @param points	list of Datapoint objects.
	 * @param IDs		list of all existing IDs.
	 * @return result	list of Trajectory-objects.
	 *************************************************/
	public static ArrayList<Trajectory> sortedDPoints(ArrayList<Datapoint> points, ArrayList<String> IDs){
		ArrayList<Trajectory> result = new ArrayList<Trajectory>();
		ArrayList<Datapoint> trajec = new ArrayList<Datapoint>();
		String currentID = points.get(0).getID();
		for(Datapoint point: points){
			if(point.getID().equals(currentID)){
				trajec.add(point);
			}
			else{
				Trajectory temp = new Trajectory(trajec);
				result.add(temp);
				trajec = new ArrayList<Datapoint>();
				trajec.add(point);
				currentID = point.getID();
			}
		}
		Trajectory temp = new Trajectory(trajec);
		result.add(temp);
		return result;
	}
	
	
	/*****************************************************************************
	 * Linear interpolation of datapoint speeds.
	 * @param trajec 			ArrayList of of trajectories.
	 * @return trajectories		Linearly interpolated ArrayList of trajectories. 
	 *****************************************************************************/
	public static ArrayList<Trajectory> linearInterpolate(ArrayList<Trajectory> trajec){
		ArrayList<Trajectory> trajectories = new ArrayList<Trajectory>();
		for(Trajectory traj: trajec){
			ArrayList<Datapoint> newPoints = new ArrayList<Datapoint>();
			ArrayList<Datapoint> points = traj.getTrajectory();
			for(int i=0; i<points.size()-1; i++){
				Datapoint d1 = new Datapoint(points.get(i));
				Datapoint d2 = new Datapoint(points.get(i+1));
				double dt = d2.getT()-d1.getT();
				double dx = d2.getX()-d1.getX();
				double dy = d2.getY()-d1.getY();
				double dz = d2.getZ()-d1.getZ();
				d1.setVx(dx/dt);
				d1.setVy(dy/dt);
				d1.setVz(dz/dt);
				newPoints.add(d1);
			}
			trajectories.add(new Trajectory(newPoints));
		}
		return trajectories;
	}
	

	/*********************************************************
	 * Time sorting method for all Datapoints.
	 * @param trajectories	list of trajectories.
	 * @param times			list of time specs.
	 * @return result		time sorted lists of Datapoints.
	 *********************************************************/
	public static ArrayList<ArrayList<Datapoint>> timeSort(ArrayList<Trajectory> trajectories, ArrayList<Double> times){
		ArrayList<ArrayList<Datapoint>> result = new ArrayList<ArrayList<Datapoint>>();
		for(int i=0; i<times.size(); i++){
			result.add(new ArrayList<Datapoint>());
		}
		for(Trajectory trajec: trajectories){
			for(Datapoint point: trajec.getTrajectory()){
				result.get(times.indexOf(point.getT())).add(point);
			}
		}
		return result;
	}
	
}
